### PR TITLE
Use Async functions for OpenAi functions

### PR DESCRIPTION
**Pull request by nandorsoma**

### CODE REVIEW

File: sqrl-openai/src/main/java/com/datasqrl/openai/FunctionExecutor.java [hunk 1]
```diff
@@ -4,32 +4,44 @@
...
 import static com.datasqrl.openai.RetryUtil.executeWithRetry;
 
 public class FunctionExecutor {
 
     private final FunctionMetricTracker metricTracker;
+    private final ExecutorService executorService;
 
     public FunctionExecutor(FunctionContext context, String functionName) {
         this.metricTracker = new FunctionMetricTracker(context, functionName);
+        this.executorService = Executors.newFixedThreadPool(10);
```
Comment (nandorsoma) [inline sqrl-openai/src/main/java/com/datasqrl/openai/FunctionExecutor.java:21]: Unfortunately, I couldn’t find a way to share the executor. The most optimal approach would be to borrow one from the context, but as far as I know, such a feature doesn’t exist. Would it make sense to make the pool size configurable via an environment variable, specific to the function? For instance, something like FUNCTION_NAME.toUpperCase() + "_POOL_SIZE".

Comment (mbroecheler) [reply sqrl-openai/src/main/java/com/datasqrl/openai/FunctionExecutor.java:21]: That's fine, we can have one threadpool per function and use a single environment variable ASYNC_FUNCTION_THREAD_POOL_SIZE to control it across all functions. We can later decide if we need to make it function specific.

File: sqrl-openai/src/main/java/com/datasqrl/openai/FunctionExecutor.java [hunk 2]
```diff
@@ -4,32 +4,44 @@
...
-
-        return ret;
+    public <T> CompletableFuture<T> executeAsync(Callable<T> task) {
+        final CompletableFuture<T> future = new CompletableFuture<>();
+        executorService.submit(() -> {
+            try {
+                metricTracker.increaseCallCount();
+                final long start = System.nanoTime();
+
+                T result = executeWithRetry(
```
Comment (mbroecheler) [inline sqrl-openai/src/main/java/com/datasqrl/openai/FunctionExecutor.java:31]: Flink async functions have retry build-in now:
https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/dev/table/config/#table-exec-async-scalar-retry-strategy

Please remove the retry logic and just keep the metrics piece.

**Review by mbroecheler — CHANGES_REQUESTED** [review]
Nice! Two smallish changes, please

**Review by mbroecheler — APPROVED** [review]

**Pull request merged**